# Week-1 Assignment

*Welcome to your first assignment for the SimuTech Winter Project 2022! I hope you are excited to implement and test everything you have learned up until now. There is an interesting set of questions for you to refine your acquired skills as you delve into hands-on coding and deepen your understanding of numpy, pandas, and data visualization libraries.*

# Section 0 : Importing Libraries

*Let's begin by importing numpy, pandas and matplotlib.*

In [2]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as pl

# Section 1 : Playing with Python and Numpy

### Q1. Matrix Multiplication

##### (i) Check if matrix multiplication is valid

In [3]:
def isValid(A,B):
    if (A.shape[1] == B.shape[0]):
        return True
    else:
        return False

##### (ii) Using loops (without using numpy)

In [4]:
def matrix_multiply(A,B):
    result=np.zeros((A.shape[0],B.shape[1]),dtype=int)
    for i in range(0,A.shape[0]):
        for j in range(0,B.shape[1]):
            for k in range(0,A.shape[1]):
                result[i,j]+=A[i,k]*B[k,j]
    return result

##### (iii) Using numpy

In [5]:
def matrix_multiply_2(A,B):
    result=np.matmul(A,B)
    return result

##### (iv) Testing your code

Run the following cell to check if your functions are working properly.

*Expected output:*
[ [102 108 114]
 [246 261 276]
 [390 414 438]
 [534 567 600] ]

In [6]:
A = np.array([
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9],
    [10, 11, 12]
])

B = np.array([
    [13, 14, 15],
    [16, 17, 18],
    [19, 20, 21]
])

if isValid(A,B):
  print(f"Result using loops: {matrix_multiply(A,B)}")
  print(f"Result using numpy: {matrix_multiply_2(A,B)}")
else:
  print(f"Matrix multiplication is not valid")

Result using loops: [[102 108 114]
 [246 261 276]
 [390 414 438]
 [534 567 600]]
Result using numpy: [[102 108 114]
 [246 261 276]
 [390 414 438]
 [534 567 600]]


### Q2. Z-Score Normalisation

Z-score normalization refers to the process of normalizing every value in a dataset such that the mean of all of the values is 0 and the standard deviation is 1.

We use the following formula to perform a z-score normalization on every value in a dataset:

New value = (x – μ) / σ

where:

x: Original value

μ: Mean of data

σ: Standard deviation of data

##### (i) Without using numpy

In [7]:
def mean(x):
    m=0
    for i in range(0,len(x)):
        m+=x[i]
    return m/len(x)

In [8]:
def standard_deviation(x):
    s=0
    for i in range(0,len(x)):
        s+=(x[i]-mean(x))**2
    std=(s/len(x))**0.5
    return std

In [9]:
def zscore_normalisation(x):
    b=np.zeros((len(x)))
    for i in range(0,len(x)):
        b[i]=(x[i]-mean(x))/standard_deviation(x)
    return b

##### (ii) Using numpy

Numpy has in_built functions for calculating mean and standard deviation

In [10]:
def zscore_normalisation_2(x):
    b=np.zeros((len(x)))
    m=np.mean(x)
    s=np.std(x)
    for i in range(0,len(x)):
        b[i]=(x[i]-m)/s
    return b

##### (iii) Testing your code

Run the following cell to check if your functions are working properly.

*Expected Output:* [-1.06753267 -0.99745394 -0.99745394 -0.81057732 -0.41346451 -0.06307086
  0.31068237  0.91803138  1.22170588  1.89913361]

In [11]:
x = [4, 7, 7, 15, 32, 47, 63, 89, 102, 131]
print(f"Result without using numpy: {zscore_normalisation(x)}")
print(f"Result using numpy: {zscore_normalisation_2(x)}")

Result without using numpy: [-1.06753267 -0.99745394 -0.99745394 -0.81057732 -0.41346451 -0.06307086
  0.31068237  0.91803138  1.22170588  1.89913361]
Result using numpy: [-1.06753267 -0.99745394 -0.99745394 -0.81057732 -0.41346451 -0.06307086
  0.31068237  0.91803138  1.22170588  1.89913361]


### Q3. Sigmoid fn and its derivative

The sigmoid function is a mathematical function that maps any input value to a value between 0 and 1.

It is defined mathematically as s(x) = 1/(1+e^(-x)).

##### (i) Write a fn to implement sigmoid fn

In [12]:
def sigmoidfn(x):
    b=1/(1+np.exp(-x))
    return b

##### (ii) Write a fn to implement derivative of sigmoid fn

In [13]:
def derivative(x):
    b=sigmoidfn(x)-(sigmoidfn(x))**2
    return b

##### (iii) Test your code

Run the following cell to check if your functions are working properly.

*Expected output:*

x on applying sigmoid activation fn is: [ [0.99987661 0.88079708 0.99330715 0.5        0.5       ]
 [0.99908895 0.99330715 0.5        0.5        0.5       ] ]

x on applying derivative of sigmoid activation fn is: [ [-1.23379350e-04 -1.04993585e-01 -6.64805667e-03 -2.50000000e-01
  -2.50000000e-01]
 [-9.10221180e-04 -6.64805667e-03 -2.50000000e-01 -2.50000000e-01
  -2.50000000e-01] ]

In [14]:
x = np.array([
    [9,2,5,0,0],
    [7,5,0,0,0]
])
print(f"x on applying sigmoid activation fn is: {sigmoidfn(x)}")
print(f"x on applying derivative of sigmoid activation fn is: {derivative(x)}")

x on applying sigmoid activation fn is: [[0.99987661 0.88079708 0.99330715 0.5        0.5       ]
 [0.99908895 0.99330715 0.5        0.5        0.5       ]]
x on applying derivative of sigmoid activation fn is: [[1.23379350e-04 1.04993585e-01 6.64805667e-03 2.50000000e-01
  2.50000000e-01]
 [9.10221180e-04 6.64805667e-03 2.50000000e-01 2.50000000e-01
  2.50000000e-01]]


# Section 2: Exploring Pandas

*You have been provided with a dataset which includes information about properties of superheated vapor.*

*The dataset consists of the thermophysical properties: specific volume, specific internal energy, specific enthalpy, specific entropy of superheated vapor.*

*Pressure is in kPa and Temperature in centigrade. In the dataframe 75, 100, 125, etc. are temperatures.*

### Read the csv file


In [19]:
df=pd.read_csv("superheated_vapor_properties.csv")

### Display the shape of data frame


In [20]:
print (df.shape)

(543, 37)


### Return an array containing names of all the columns

In [21]:
print (df.columns)

Index(['Pressure', 'Property', 'Liq_Sat', 'Vap_Sat', '75', '100', '125', '150',
       '175', '200', '220', '225', '240', '250', '260', '275', '280', '290',
       '300', '320', '325', '340', '350', '360', '375', '380', '400', '425',
       '450', '475', '500', '525', '550', '575', '600', '625', '650'],
      dtype='object')


### Display the number of null values in each column of the dataframe



In [22]:
print (df.isnull().sum())


Pressure      0
Property      0
Liq_Sat       0
Vap_Sat       0
75          527
100         507
125         487
150         447
175         383
200         319
220         407
225         399
240         407
250         343
260         383
275         339
280         379
290         487
300          59
320         479
325         135
340         476
350          67
360         476
375         203
380         476
400           0
425         204
450           0
475         204
500           0
525         272
550           0
575         340
600           0
625         476
650           0
dtype: int64


### Create a column which contains the Pressure and Property columns, seperated with 'at' (For eg. V at 1, H at 101.325). Using this print the following:
- Enthalpy at 75 kPa and 573 K
- Entropy at 493 K and 250 kPa



In [29]:
df=df.astype(str)
df["Property and Pressure"]=df["Property"] + " at " + df["Pressure"]
print (df)

    Pressure Property Liq_Sat   Vap_Sat        75       100       125  \
0        1.0        V     1.0  129200.0  160640.0  172180.0  183720.0   
1        1.0        U  29.334    2385.2    2480.8    2516.4    2552.3   
2        1.0        H  29.335    2514.4    2641.5    2688.6    2736.0   
3        1.0        S   0.106    8.9767    9.3828    9.5136    9.6365   
4       10.0        V    1.01   14670.0   16030.0   17190.0   18350.0   
..       ...      ...     ...       ...       ...       ...       ...   
538  11200.0        H  1458.9    2705.4       nan       nan       nan   
539  11200.0        S   3.444    5.5476       nan       nan       nan   
540  11400.0        V   1.504    15.284       nan       nan       nan   
541  11400.0        U  1450.0    2527.2       nan       nan       nan   
542  11400.0        H  1467.2    2701.5       nan       nan       nan   

          150       175       200  ...       450     475       500     525  \
0    195270.0  206810.0  218350.0  ...  33373

### Find out the column with the highest number of missing values

In [24]:
a=df.isnull().sum()
print (a.index[a==a.max()])


Index(['75'], dtype='object')


### What is the average enthalpy of Sat. Liq. at all different pressures in the dataset?

In [25]:
H_data=df[df["Property"]=="H"]
print (H_data["Liq_Sat"].mean())


936.9707720588235


### Separate out the V,U,H,S data from the dataset into V_data, U_data, H_data, S_data

In [26]:
V_data=df[df["Property"]=="V"]
H_data=df[df["Property"]=="H"]
S_data=df[df["Property"]=="S"]
U_data=df[df["Property"]=="U"]

# Section 4 : Conculsion




*Congratulations on reaching this point! I hope you had fun solving your first assignment and have also built confidence in applying these libraries. If you are wondering, we will cover more about z-score normalization in Week 2, and the sigmoid function will be used in Week 3. After completing this assignment, you are now prepared to learn about machine learning techniques and implement your own machine learning models.*